lib/ecto/discrete_range.ex

```diff
@@ -0,0 +1,311 @@
+  """
+  @type allens_relation ::
+          :equals | :finishes | :starts | :meets | :before | :after | :during | :overlaps
+
+  @doc """
+  Creates a new interval.
+
+    * `:succ_fn` computes the next value;
+    * `:comp_fn` comparator for the given type;
+    * `:lower` the lower term of the interval;
+    * `:upper` the upper term of the interval;
+    * `:mode` defines how to interpret interval boundaries;
+  """
+  @spec new(succ_fn, comp_fn, Ecto.Type, t, t, {:open | :closed, :open | :closed}) ::
+          {:ok, {t, t}} | :error
+  def new(succ_fn, comp_fn, ecto_ty, lower, upper, mode) do
+    interval_fn =
+      case mode do
+        {:closed, :open} ->
+          fn a, b -> {:ok, {a, b}} end
+
+        {:closed, :closed} ->
+          fn a, b -> {:ok, {a, succ_term(succ_fn, b)}} end
+
+        {:open, :open} ->
+          fn a, b -> {:ok, {succ_term(succ_fn, a), b}} end
+
+        {:open, :closed} ->
+          fn a, b -> {:ok, {succ_term(succ_fn, a), succ_term(succ_fn, b)}} end
+
+        _ ->
+          fn _a, _b -> :error end
+      end
+
+    with {:ok, lower} <- cast_term(:l, ecto_ty, lower),
+         {:ok, upper} <- cast_term(:r, ecto_ty, upper),
+         {:ok, {lower, upper}} <- interval_fn.(lower, upper),
+         true <- :gt != comp_term(comp_fn, lower, upper) do
+      {:ok, {lower, upper}}
+    else
+      _ -> :error
+    end
+  end
+
+  @spec load(succ_fn, Ecto.Type.t(), %Postgrex.Range{}) :: {:ok, {t, t}} | :error
+  def load(succ_fn, ecto_ty, value = %Postgrex.Range{}) do
+    with {:ok, lower} <- load_term(:l, ecto_ty, value.lower),
+         {:ok, upper} <- load_term(:u, ecto_ty, value.upper) do
+      lower =
+        if value.lower_inclusive do
+          lower
+        else
+          succ_term(succ_fn, lower)
+        end
+
+      upper =
+        if value.upper_inclusive do
+          succ_term(succ_fn, upper)
+        else
+          upper
+        end
+
+      {:ok, {lower, upper}}
+    end
+  end
+
+  @spec cast(succ_fn, comp_fn, Ecto.Type.t(), String.t()) :: {:ok, {t, t}} | :error
+  def cast(succ_fn, comp_fn, ecto_ty, value) when is_binary(value) do
+    with [lower, upper] <- String.split(value, ",", parts: 2),
+         {:ok, lmode} <- cast_mode(:l, String.at(lower, 0)),
+         {:ok, rmode} <- cast_mode(:u, String.at(upper, -1)),
+         {:ok, lower} <- cast_term(:l, ecto_ty, String.slice(lower, 1..-1)),
+         {:ok, upper} <- cast_term(:u, ecto_ty, String.slice(upper, 0..-2)) do
+      new(succ_fn, comp_fn, ecto_ty, lower, upper, {lmode, rmode})
+    else
+      _ -> :error
+    end
+  end
+
+  @spec dump(Ecto.Type.t(), t, t) :: {:ok, Postgrex.Range.t()} | :error
+  def dump(ecto_ty, lower, upper) do
+    with {:ok, lower} <- dump_term(ecto_ty, lower),
+         {:ok, upper} <- dump_term(ecto_ty, upper) do
+      {:ok,
+       %Postgrex.Range{
+         lower: lower,
+         upper: upper,
+         lower_inclusive: lower != nil,
+         upper_inclusive: false
+       }}
```

Comment:
This will make the code not compile if the application uses ecto with any driver besides postgrex. To be viable, there should be a transparent way to create ranges in Ecto that is then passed to the driver

Reply:
i've avoided using `if Code.ensure_loaded?(Postgrex)` like ecto does in the adapters modules at this point because (a) i don't know if that is acceptable here (b) i'd like to probe if something like this is wanted in the ecto library before doing any more work.

is that `if` the way to go here?